File: Code/139.word-break.py
import copy
#
# @lc app=leetcode id=139 lang=python3
#
# [139] Word Break
#

# @lc code=start
class Solution:
    words = []

    def isEqual(self, s: str):
        wd = ""
        for word in words:
            wd += word
        return wd == s  

    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        if s in wordDict:
            words.append(s)
            return True
        for i in range(len(s)):
            if s[0:i+1] in wordDict:
                words.append(s[0:i+1])
                if wordBreak(s[i+1:], wordDict) != True:
                    words.clear()
            if self.isEqual(s):
                return True
        return False
     
         
# @lc code=end

File: Code/139.word-break.java
import java.util.*;
/*
 * @lc app=leetcode id=139 lang=java
 *
 * [139] Word Break
 */

// @lc code=start
class Solution {
    String str;
    Set<Integer> cache;
    List<String> list;
    public boolean wordBreak(String s, List<String> wordDict) {
       str=s;
       cache=new HashSet<Integer>();
       list=wordDict;
       return wordBreak(0);
    }
     private boolean wordBreak(int d){
        if(d == str.length()) return true;
        if(cache.contains(d)) return false;//无需再次重复计算，剪枝
        for(String word : list){
            if(str.startsWith(word, d)){
                if(wordBreak(d+word.length())) return true;
                cache.add(d+word.length());//已经为false的
            }
        }
        return false;
    }
}
// @lc code=end

